internal/pkg/opentelemetry/controller_test.go

```diff
@@ -390,3 +362,38 @@ func TestShutdown(t *testing.T) {
 	assert.True(t, exporter.called, "Exporter not shutdown")
 	assert.Equal(t, uint32(nSpan), exporter.exported.Load(), "Pending spans not flushed")
 }
+
+func TestControllerTraceConcurrentSafe(t *testing.T) {
+	exporter := tracetest.NewInMemoryExporter()
```

Comment:
It might be an ignorant question: How do we validate in this test that it behaves well concurrently? We create `goroutines` spans but should we make sure they are created as expected with some asserts?

Another question is whether the controller's receivers have concurrency guarantees or not. (assuming we are planning to replace the controller)

Reply:
It's a race test so it doesn't need to have direct failure cases, the failure will be a data race.

For example this is what happens if you run this on `main`:

```terminal
$ go test -race
==================
WARNING: DATA RACE
Write at 0x00c00019c0c0 by goroutine 28:
  runtime.mapassign()
      /usr/lib/go/src/runtime/map.go:615 +0x0
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.(*Controller).getTracer()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller.go:32 +0x32b
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.(*Controller).Trace()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller.go:46 +0x10e
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.TestControllerTraceConcurrentSafe.func2()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller_test.go:394 +0x609

Previous read at 0x00c00019c0c0 by goroutine 33:
  runtime.mapaccess2()
      /usr/lib/go/src/runtime/map.go:479 +0x0
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.(*Controller).getTracer()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller.go:29 +0xe7
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.(*Controller).Trace()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller.go:46 +0x10e
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.TestControllerTraceConcurrentSafe.func2()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller_test.go:394 +0x609

Goroutine 28 (running) created at:
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.TestControllerTraceConcurrentSafe()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller_test.go:386 +0x44b
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /usr/lib/go/src/testing/testing.go:1743 +0x44

Goroutine 33 (finished) created at:
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.TestControllerTraceConcurrentSafe()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller_test.go:386 +0x44b
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /usr/lib/go/src/testing/testing.go:1743 +0x44
==================
==================
WARNING: DATA RACE
Write at 0x00c0001e0190 by goroutine 28:
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.(*Controller).getTracer()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller.go:32 +0x33a
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.(*Controller).Trace()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller.go:46 +0x10e
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.TestControllerTraceConcurrentSafe.func2()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller_test.go:394 +0x609

Previous write at 0x00c0001e0190 by goroutine 33:
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.(*Controller).getTracer()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller.go:32 +0x33a
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.(*Controller).Trace()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller.go:46 +0x10e
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.TestControllerTraceConcurrentSafe.func2()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller_test.go:394 +0x609

Goroutine 28 (running) created at:
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.TestControllerTraceConcurrentSafe()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller_test.go:386 +0x44b
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /usr/lib/go/src/testing/testing.go:1743 +0x44

Goroutine 33 (finished) created at:
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.TestControllerTraceConcurrentSafe()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller_test.go:386 +0x44b
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /usr/lib/go/src/testing/testing.go:1743 +0x44
==================
==================
WARNING: DATA RACE
Write at 0x00c00019c0c0 by goroutine 30:
  runtime.mapassign()
      /usr/lib/go/src/runtime/map.go:615 +0x0
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.(*Controller).getTracer()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller.go:32 +0x32b
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.(*Controller).Trace()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller.go:46 +0x10e
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.TestControllerTraceConcurrentSafe.func2()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller_test.go:394 +0x609

Previous write at 0x00c00019c0c0 by goroutine 31:
  runtime.mapassign()
      /usr/lib/go/src/runtime/map.go:615 +0x0
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.(*Controller).getTracer()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller.go:32 +0x32b
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.(*Controller).Trace()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller.go:46 +0x10e
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.TestControllerTraceConcurrentSafe.func2()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller_test.go:394 +0x609

Goroutine 30 (running) created at:
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.TestControllerTraceConcurrentSafe()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller_test.go:386 +0x44b
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /usr/lib/go/src/testing/testing.go:1743 +0x44

Goroutine 31 (finished) created at:
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.TestControllerTraceConcurrentSafe()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller_test.go:386 +0x44b
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /usr/lib/go/src/testing/testing.go:1743 +0x44
==================
==================
WARNING: DATA RACE
Read at 0x00c0001e01c0 by goroutine 25:
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.(*Controller).getTracer()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller.go:29 +0xf8
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.(*Controller).Trace()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller.go:46 +0x10e
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.TestControllerTraceConcurrentSafe.func2()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller_test.go:394 +0x609

Previous write at 0x00c0001e01c0 by goroutine 30:
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.(*Controller).getTracer()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller.go:32 +0x33a
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.(*Controller).Trace()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller.go:46 +0x10e
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.TestControllerTraceConcurrentSafe.func2()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller_test.go:394 +0x609

Goroutine 25 (running) created at:
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.TestControllerTraceConcurrentSafe()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller_test.go:386 +0x44b
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /usr/lib/go/src/testing/testing.go:1743 +0x44

Goroutine 30 (finished) created at:
  go.opentelemetry.io/auto/internal/pkg/opentelemetry.TestControllerTraceConcurrentSafe()
      /home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/opentelemetry/controller_test.go:386 +0x44b
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /usr/lib/go/src/testing/testing.go:1743 +0x44
==================
--- FAIL: TestControllerTraceConcurrentSafe (0.00s)
    testing.go:1399: race detected during execution of test
FAIL
```

Reply:
> Another question is whether the collector's receivers have concurrency guarantees or not. (assuming we are planning to replace the controller)

Switching to the will require us to build a `Receiver` type that gets data from the auto-instrumentation. It will be up to us to ensure that type's method for receiving data is concurrent safe. There is nothing preventing our ability to build a concurrent-safe method based on existing PoCs.

Reply:
https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/1249 is looking to add a GitHub action workflow that will automate this testing FWIW.

I needs to also fix other data-races we have in our testing so it is in a separate PR.